python/cudf/cudf/core/dataframe.py

```diff
@@ -284,12 +287,18 @@ def _init_from_dict_like(self, data, index=None, columns=None):
             # not in `columns`
             keys = [key for key in data.keys() if key in columns]
             data = {key: data[key] for key in keys}
+            extra_cols = [col for col in columns if col not in data.keys()]
             if keys:
                 # if keys is non-empty,
                 # add null columns for all values
                 # in `columns` that don't exist in `keys`:
-                extra_cols = [col for col in columns if col not in data.keys()]
                 data.update({key: None for key in extra_cols})
+            else:
+                # if keys is empty,
+                # it means that none of the actual keys in `data`
+                # matches `columns` hence only update `data`
+                # with `columns` as keys with empty columns.
+                data.update({key: [] for key in extra_cols})
```

Comment:
Could we use `column_empty` instead of `[]` here to avoid the HtoD overhead?

Reply:
I may be mistaken, but it looks like we can do instead:

```python
data = dict.fromkeys(['a', 'b'], cudf.core.column.column_empty(0))
```

If so, the comments above may be misleading, as we're not "updating" `data`, but rather setting it to have `extra_cols` as keys and empty columns as values.

Reply:
Made this change 👍  @shwina